flink-ml-iteration/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/DataCache.java

```diff
@@ -0,0 +1,351 @@
+
+    private final SupplierWithException<Path, IOException> pathGenerator;
+
+    private final MemorySegmentPool segmentPool;
+
+    private final List<Segment> finishedSegments;
+
+    private SegmentWriter<T> currentWriter;
+
+    public DataCache(
+            TypeSerializer<T> serializer,
+            FileSystem fileSystem,
+            SupplierWithException<Path, IOException> pathGenerator)
+            throws IOException {
+        this(serializer, fileSystem, pathGenerator, null, Collections.emptyList());
+    }
+
+    public DataCache(
+            TypeSerializer<T> serializer,
+            FileSystem fileSystem,
+            SupplierWithException<Path, IOException> pathGenerator,
+            MemorySegmentPool segmentPool)
+            throws IOException {
+        this(serializer, fileSystem, pathGenerator, segmentPool, Collections.emptyList());
+    }
+
+    public DataCache(
+            TypeSerializer<T> serializer,
+            FileSystem fileSystem,
+            SupplierWithException<Path, IOException> pathGenerator,
+            MemorySegmentPool segmentPool,
+            List<Segment> finishedSegments)
+            throws IOException {
+        this.fileSystem = fileSystem;
+        this.pathGenerator = pathGenerator;
+        this.segmentPool = segmentPool;
+        this.serializer = serializer;
+        this.finishedSegments = new ArrayList<>();
+        this.finishedSegments.addAll(finishedSegments);
+        for (Segment segment : finishedSegments) {
+            tryCacheSegmentToMemory(segment);
+        }
+        this.currentWriter = createSegmentWriter();
+    }
+
+    public void addRecord(T record) throws IOException {
+        try {
+            currentWriter.addRecord(record);
+        } catch (SegmentNoVacancyException e) {
+            currentWriter.finish().ifPresent(finishedSegments::add);
+            currentWriter = new FileSegmentWriter<>(serializer, pathGenerator.get());
+            currentWriter.addRecord(record);
+        }
+    }
+
+    /** Finishes adding records and closes resources occupied for adding records. */
+    public void finish() throws IOException {
+        if (currentWriter == null) {
+            return;
+        }
+
+        currentWriter.finish().ifPresent(finishedSegments::add);
+        currentWriter = null;
+    }
+
+    /** Cleans up all previously added records. */
+    public void cleanup() throws IOException {
+        finishCurrentSegmentIfAny();
+        for (Segment segment : finishedSegments) {
+            if (segment.getFileSegment() != null) {
+                fileSystem.delete(segment.getFileSegment().getPath(), false);
+            }
+            if (segment.getMemorySegment() != null) {
+                segmentPool.returnAll(segment.getMemorySegment().getCache());
+            }
+        }
+        finishedSegments.clear();
+    }
+
+    private void finishCurrentSegmentIfAny() throws IOException {
+        if (currentWriter == null || currentWriter.getCount() == 0) {
```

Comment:
It is possible. For example in `ReplayOperator`, `finish()` is invoked at `onEpochWatermarkIncrement` while `finishCurrentSegmentIfAny()` is invoked at `snapshotState`. Since `snapshotState` can happen before or after `onEpochWatermarkIncrement`, `finish()` and `finishCurrentSegmentIfAny()` might be invoked repeatedly in any order, and the `if` condition if unavoidable. If we remove the `if` from DataCacheWriter, users still need to add it in their code.